src/nnet3/nnet-simple-component.cc

```diff
@@ -4261,6 +4254,273 @@ std::string MaxpoolingComponent::Info() const {
+  KALDI_ASSERT((input_z_dim_ - pool_z_size_) % pool_z_step_  == 0);
+}
+
+// initialize the component using configuration file
+void Maxpooling3dComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 input_x_dim = -1, input_y_dim = -1, input_z_dim = -1;
+  int32 pool_x_size = -1, pool_y_size = -1, pool_z_size = -1;
+  int32 pool_x_step = -1, pool_y_step = -1, pool_z_step = -1;
+  bool ok = true;
+
+  ok = ok && cfl->GetValue("input-x-dim", &input_x_dim);
+  ok = ok && cfl->GetValue("input-y-dim", &input_y_dim);
+  ok = ok && cfl->GetValue("input-z-dim", &input_z_dim);
+  ok = ok && cfl->GetValue("pool-x-size", &pool_x_size);
+  ok = ok && cfl->GetValue("pool-y-size", &pool_y_size);
+  ok = ok && cfl->GetValue("pool-z-size", &pool_z_size);
+  ok = ok && cfl->GetValue("pool-x-step", &pool_x_step);
+  ok = ok && cfl->GetValue("pool-y-step", &pool_y_step);
+  ok = ok && cfl->GetValue("pool-z-step", &pool_z_step);
+
+  if (cfl->HasUnusedValues())
+    KALDI_ERR << "Could not process these elements in initializer: "
+              << cfl->UnusedValues();
+  if (!ok)
+    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+  Init(input_x_dim, input_y_dim, input_z_dim,
+       pool_x_size, pool_y_size, pool_z_size,
+       pool_x_step, pool_y_step, pool_z_step);
+}
+
+// Method to convert from a matrix representing a minibatch of vectorized
+// 3D tensors to patches for 3d max pooling, each patch corresponds to
+// one pool output dot product in the pooling
+void Maxpooling3dComponent::InputToInputPatches(
+    const CuMatrixBase<BaseFloat>& in,
+    CuMatrix<BaseFloat> *patches) const{
+  int32 num_pools_x = 1 + (input_x_dim_ - pool_x_size_) / pool_x_step_;
+  int32 num_pools_y = 1 + (input_y_dim_ - pool_y_size_) / pool_y_step_;
+  int32 num_pools_z = 1 + (input_z_dim_ - pool_z_size_) / pool_z_step_;
+
+  std::vector<int32> column_map(patches->NumCols());
+  int32 column_map_size = column_map.size();
+  for (int32 x = 0, index =0; x < pool_x_size_; x++) {
+    for (int32 y = 0; y < pool_y_size_; y++) {
+      for (int32 z = 0; z < pool_z_size_; z++) {
+        // given the local node coordinate, group them from each pool
+        for (int32 x_pool = 0; x_pool < num_pools_x; x_pool++) {
+          for (int32 y_pool = 0; y_pool < num_pools_y; y_pool++) {
+            for (int32 z_pool = 0; z_pool < num_pools_z; z_pool++, index++) {
+              KALDI_ASSERT(index < column_map_size);
+              column_map[index] = (x_pool * pool_x_step_ + x) * input_y_dim_ * input_z_dim_ +
+                                  (y_pool * pool_y_step_ + y) * input_z_dim_ +
+                                  (z_pool * pool_z_step_ + z);
+
+            }
+          }
+        }
+      }
+    }
+  }
+  CuArray<int32> cu_cols(column_map);
+  patches->CopyCols(in, cu_cols);
+}
+
+/*
+  3d max pooling propagate function
+  It is assumed that each row of the input matrix
+  is a vectorized 3D-tensor in zxy format.
```

Comment:
is a vectorized 3D-tensor of TensorVectorizationType zxy (see ConvolutionComponent description in nnet-simple-component.h for more information).